### PR TITLE
ILE-4 added project backend

### PR DESCRIPTION
**Pull request by CForde-Kainos**

### CODE REVIEW

**Review by dana-kainos — APPROVED** [review]

File: src/main/java/org/example/controllers/ProjectController.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        }
+    }
+
+    @POST
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response createProject(
+            final ProjectRequest project
+    ) {
+        try {
+            return Response.ok().entity(
```
Comment (lukemKainos) [inline src/main/java/org/example/controllers/ProjectController.java]: perhaps instead of ok it should be created?

File: src/main/java/org/example/daos/ProjectDao.java [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+            ResultSet rs = st.getGeneratedKeys();
+            if (rs.next()) {
+                return rs.getInt(1);
+            }
+        }
+
+        return -1;
+    }
+
+    public void updateSalesEmployee(
```
Comment (lukemKainos) [inline src/main/java/org/example/daos/ProjectDao.java]: this function updates project no?

File: src/main/java/org/example/models/Project.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
+package org.example.models;
+
+import java.sql.Date;
+
+public class Project {
+    private int projectId;
+    private String name;
+    private float value;
+    private int techleadId;
```
Comment (lukemKainos) [inline src/main/java/org/example/models/Project.java]: this isnt camel case

File: src/main/java/org/example/models/ProjectRequest.java [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import java.sql.Date;
+
+public class ProjectRequest {
+    private String name;
+    private float value;
+    private int techleadId;
```
Comment (lukemKainos) [inline src/main/java/org/example/models/ProjectRequest.java]: camel case

File: src/main/java/org/example/models/ProjectRequest.java [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    private int clientId;
+    private int salesEmployeeId;
+    private Date startDate;
+    private Date finishDate;
+    private float commissionRate;
+    @JsonCreator
+    public ProjectRequest(
+            @JsonProperty("projectId") final int projectId,
```
Comment (lukemKainos) [inline src/main/java/org/example/models/ProjectRequest.java]: this isnt used

File: src/main/java/org/example/services/ProjectService.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    }
+
+    public List<Project> getProjects()
+            throws SQLException {
+        return projectDao.getProjects();
+    }
+
+    public Project getProject(int id) throws SQLException,
+            DoesNotExistException {
+        Project salesEmployee = projectDao.getProjectById(id);
```
Comment (lukemKainos) [inline src/main/java/org/example/services/ProjectService.java]: how is this a project?

File: src/main/java/org/example/services/ProjectService.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    public int createProject(ProjectRequest projectRequest)
+            throws SQLException, FailedToCreateException {
+        int salesEmployeeId = projectDao.createProject(projectRequest);
+        if (salesEmployeeId == -1) { throw new FailedToCreateException(Entity.PROJECT); }
+        return salesEmployeeId;
+    }
+
+    public void deleteProject(int id) throws SQLException, DoesNotExistException {
+        Project salesEmployee = projectDao.getProjectById(id);
+        if (salesEmployee == null) { throw new DoesNotExistException(Entity.PROJECT); }
```
Comment (lukemKainos) [inline src/main/java/org/example/services/ProjectService.java]: this code is duplicated multiple times

**Review by lukemKainos — CHANGES_REQUESTED** [review]

File: src/main/java/org/example/daos/ProjectDao.java [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+public class ProjectDao {
+    private enum Indexes {
+        name(1),
+        value(2),
+        techLeadId(3),
+        clientId(4),
+        salesEmployeeId(5),
+        startDate(6),
+        finishDate(7),
+        updateId(8);
```
Comment (lukemKainos) [inline src/main/java/org/example/daos/ProjectDao.java]: enums should be formatted in upper case as they are effectively private static variables

**Review by lukemKainos — CHANGES_REQUESTED** [review]

File: src/main/java/org/example/TestApplication.java [hunk 9]
```diff
@@ -7,11 +7,13 @@
 import io.federecio.dropwizard.swagger.SwaggerBundleConfiguration;
 import org.example.controllers.SalesEmployeeController;
 import org.example.controllers.DeliveryEmployeeController;
+import org.example.controllers.ProjectController;
 import org.example.controllers.TestController;
 import org.example.daos.DeliveryEmployeeDao;
-import org.example.daos.SalesEmployeeDao;
```
Comment (declanlarkin) [inline src/main/java/org/example/TestApplication.java:27]: Why is this being removed? It is still used

File: src/main/java/org/example/daos/ProjectDao.java [hunk 10]
```diff
@@ -0,0 +1,168 @@
+package org.example.daos;
+
+import org.example.models.DeliveryEmployee;
```
Comment (declanlarkin) [inline src/main/java/org/example/daos/ProjectDao.java]: Unused?

File: src/main/java/org/example/daos/ProjectDao.java [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+public class ProjectDao {
+    private enum Indexes {
+        NAME(1),
+        VALUE(2),
+        TECHLEADID(3),
+        CLIENTID(4),
+        SALESEMPLOYEEID(5),
+        STRATDATE(6),
+        FINISHDATE(7),
+        UPDATEID(8);
```
Comment (declanlarkin) [inline src/main/java/org/example/daos/ProjectDao.java]: camelCase please

File: src/main/java/org/example/exceptions/Entity.java [hunk 12]
```diff
@@ -1,6 +1,8 @@
 package org.example.exceptions;
 
 public enum Entity {
+    DELIVERY_EMPLOYEE("Delivery Employee"),
```
Comment (declanlarkin) [inline src/main/java/org/example/exceptions/Entity.java:4]: duplicate

File: src/main/resources/db/migration/V4_CreateProject.sql [hunk 13]
```diff
@@ -0,0 +1,13 @@
+CREATE TABLE project (
+    projectId INT NOT NULL AUTO_INCREMENT,
+    `name` VARCHAR(50) NOT NULL,
+    `value` DECIMAL(11, 2) NOT NULL,
+    techLeadId INT NOT NULL,
```
Comment (declanlarkin) [inline src/main/resources/db/migration/V4_CreateProject.sql:5]: looks like this is supposed to be a fkey that links to a techLead lkp table? Is this in scope here or coming in another ticket?

File: src/main/resources/db/migration/V4_CreateProject.sql [hunk 14]
```diff
@@ -0,0 +1,13 @@
...
+    clientId INT NOT NULL,
+    salesEmployeeId INT NOT NULL,
+    startDate DateTime NOT NULL DEFAULT CURRENT_TIMESTAMP,
+    finishDate DateTime,
```
Comment (declanlarkin) [inline src/main/resources/db/migration/V4_CreateProject.sql:9]: very pedantic but I think endDate works better here

File: src/main/java/org/example/models/ProjectRequest.java [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+    private Date finishDate;
+    @JsonCreator
+    public ProjectRequest(
+            @JsonProperty("name") final String name,
+            @JsonProperty("value") final float value,
+            @JsonProperty("techLeadId") final int techLeadId,
+            @JsonProperty("clientId") final int clientId,
+            @JsonProperty("salesEmployeeId") final int salesEmployeeId,
+            @JsonProperty("startDate") final Date startDate,
+            @JsonProperty("finishDate") final Date finishDate)
```
Comment (declanlarkin) [inline src/main/java/org/example/models/ProjectRequest.java]: For readability I would have the `final` before @JSONProperty

File: src/main/resources/db/migration/V4_CreateProject.sql (hunk 14 above)
Comment (lukemKainos) [reply src/main/resources/db/migration/V4_CreateProject.sql:9]: let leave at finishdate for now, matches prior schema

File: src/main/resources/db/migration/V4_CreateProject.sql (hunk 13 above)
Comment (CForde-Kainos) [reply src/main/resources/db/migration/V4_CreateProject.sql:5]: see line 12

File: src/main/java/org/example/daos/ProjectDao.java (hunk 11 above)
Comment (lukemKainos) [reply src/main/java/org/example/daos/ProjectDao.java]: not for an enum should be _
@CForde-Kainos  also STRATDATE

File: src/main/java/org/example/exceptions/Entity.java [hunk 16]
```diff
@@ -1,10 +1,10 @@
 package org.example.exceptions;
 
 public enum Entity {
-    USER("User"),
-    SALES_EMPLOYEE("Sales Employee"),
     DELIVERY_EMPLOYEE("Delivery Employee"),
-    CLIENT("Client");
+    PROJECT("Project")
+    SALES_EMPLOYEE("Sales Employee"),
+    DELIVERY_EMPLOYEE("Delivery Employee");
```
Comment (lukemKainos) [inline src/main/java/org/example/exceptions/Entity.java]: this is still here?

Comment (CForde-Kainos) [reply src/main/java/org/example/exceptions/Entity.java]: what is still here?

**Review by lukemKainos — APPROVED** [review]
